### PR TITLE
91367 Debt Resolution Exhausted Metrics

### PR DESCRIPTION
**Pull request by digitaldrk**
## Summary

- *This work is behind a feature toggle (flipper): NO*
- *(Summarize the changes that have been made to the platform)*: To aid in the silent errors initiative, we are adding exhaustion blocks to our jobs.
- *(If bug, how to reproduce)*: N/A
- *(What is the solution, why is this the solution?)*: We need to get more visibility into our Sidekiq Jobs to ensure veterans  get their objectives accomplished
- *(Which team do you work for, does your team own the maintenance of this component?)*:  You can find me at #debt-resolution
- *(If introducing a flipper, what is the success criteria being targeted?)* N/A

## Related issue(s)

- *Link to ticket created in va.gov-team repo OR screenshot of Jira ticket if your team uses Jira*
  - https://app.zenhub.com/workspaces/vsa---debt-607736a6c8b7e2001084e3ab/issues/gh/department-of-veterans-affairs/va.gov-team/91367
- *Link to previous change of the code/bug (if applicable)* N/A
- *Link to epic if not included in ticket*

## Testing done

- [x] *New code is covered by unit tests*
- *Describe what the old behavior was prior to the change*
- *Describe the steps required to verify your changes are working as expected. Exclusively stating 'Specs run' is NOT acceptable as appropriate testing*
- *If this work is behind a flipper:*
  - *Tests need to be written for both the flipper on and flipper off scenarios. [Docs](https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-Backendexample).*
  - *What is the testing plan for rolling out the feature?*

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
*(Describe what parts of the site are impacted and*if*code touched other areas)*

## Acceptance criteria

- [x]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x]  No error nor warning in the console.
- [x]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [x]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

**Review by JunTaoLuo — COMMENTED** [review]
Some minor comments.

File: modules/debts_api/app/sidekiq/debts_api/v0/form5655/vba_submission_job.rb [hunk 1]
```diff
@@ -13,8 +13,16 @@ class V0::Form5655::VBASubmissionJob
     class MissingUserAttributesError < StandardError; end
 
     sidekiq_retries_exhausted do |job, _ex|
+      StatsD.increment("#{STATS_KEY}.retries_exhausted")
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply modules/debts_api/app/sidekiq/debts_api/v0/form5655/vba_submission_job.rb:16]: I like the approach here so I updated my PR with it. If we're making the identical changes I think git won't need to revolve conflicts 😄

**Review by JunTaoLuo — APPROVED** [review]
LGTM

**Review by JunTaoLuo — DISMISSED** [review]

File: app/sidekiq/copay_notifications/mcp_notification_email_job.rb [hunk 2]
```diff
@@ -14,6 +14,8 @@ class McpNotificationEmailJob
     include SentryLogging
     sidekiq_options retry: 14
 
+    # Should we remove this? Will it be needed in the future?
```
[the remark this answers is not in the document]
Comment (JunTaoLuo) [reply app/sidekiq/copay_notifications/mcp_notification_email_job.rb:17]: We will remove unused code in a separate PR to avoid exceeding the PR size

**Review by JunTaoLuo — DISMISSED** [review]

File: modules/debts_api/app/workers/debts_api/v0/form5655/vha/sharepoint_submission_job.rb [hunk 3]
```diff
@@ -11,7 +11,8 @@ class V0::Form5655::VHA::SharepointSubmissionJob
     sidekiq_options retry: 4
 
     sidekiq_retries_exhausted do |job, _ex|
-      StatsD.increment("#{STATS_KEY}.failure")
+      StatsD.increment("#{STATS_KEY}.failure") # Deprecate this in favor of exhausted naming convention below
```
Comment (JunTaoLuo) [inline modules/debts_api/app/workers/debts_api/v0/form5655/vha/sharepoint_submission_job.rb:14]: Note we'll deprecate this once we have enough metrics collected using the new naming convention.

**Review by JunTaoLuo — APPROVED** [review]

**Review by Dr-Pongo — APPROVED** [review]
LGTM! 🚀

File: app/sidekiq/copay_notifications/mcp_notification_email_job.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (LindseySaari) [reply app/sidekiq/copay_notifications/mcp_notification_email_job.rb:17]: I would recommend removing this comment if it's unnecessary

File: app/sidekiq/copay_notifications/parse_new_statements_job.rb [hunk 4]
```diff
@@ -10,9 +10,19 @@ class ParseNewStatementsJob
...
     # number of jobs to perform at next interval
     BATCH_SIZE = Settings.mcp.notifications.batch_size
+    STATSD_KEY_PREFIX = 'api.copay_notifications.json_file'
+
+    sidekiq_retries_exhausted do |_msg, ex|
+      StatsD.increment("#{STATSD_KEY_PREFIX}.retries_exhausted")
+      Rails.logger.error <<~LOG
+        CopayNotifications::ParseNewStatementsJob retries exhausted:
+        Exception: #{ex.class} - #{ex.message}
+        Backtrace: #{ex.backtrace.join("\n")}
```
Comment (LindseySaari) [inline app/sidekiq/copay_notifications/parse_new_statements_job.rb:20]: Is there any PII in the backtrace?

File: modules/debts_api/app/sidekiq/debts_api/v0/form5655/vha_submission_job.rb [hunk 5]
```diff
@@ -13,8 +13,16 @@ class V0::Form5655::VHASubmissionJob
...
+      submission_id = job['args'][0]
       user_uuid = job['args'][1]
-      UserProfileAttributes.find(user_uuid)&.destroy
+      Rails.logger.error <<~LOG
+        V0::Form5655::VHASubmissionJob retries exhausted:
+        Exception: #{ex.class} - #{ex.message}
+        Backtrace: #{ex.backtrace.join("\n")}
+        submission_id: #{submission_id} | user_id: #{user_uuid}
+      LOG
+      UserProfileAttributes.find(user_uuid)&.destroy # TODO: figure out why are we destroying here.
```
Comment (LindseySaari) [inline modules/debts_api/app/sidekiq/debts_api/v0/form5655/vha_submission_job.rb:25]: I would recommend understand the "# TODO: figure out why are we destroying here." before committing this

File: modules/debts_api/app/workers/debts_api/v0/form5655/vha/vbs_submission_job.rb [hunk 6]
```diff
@@ -13,10 +13,11 @@ class V0::Form5655::VHA::VBSSubmissionJob
     class MissingUserAttributesError < StandardError; end
 
     sidekiq_retries_exhausted do |job, _ex|
-      StatsD.increment("#{STATS_KEY}.failure")
+      StatsD.increment("#{STATS_KEY}.failure") # Deprecate this in favor of exhausted naming convention below
+      StatsD.increment("#{STATS_KEY}.retries_exhausted")
       submission_id = job['args'][0]
       user_uuid = job['args'][1]
-      UserProfileAttributes.find(user_uuid)&.destroy
+      UserProfileAttributes.find(user_uuid)&.destroy # TODO: figure out why are we destroying here.
```
Comment (LindseySaari) [inline modules/debts_api/app/workers/debts_api/v0/form5655/vha/vbs_submission_job.rb:20]: Same comment here

File: app/sidekiq/copay_notifications/parse_new_statements_job.rb (hunk 4 above)
Comment (JunTaoLuo) [reply app/sidekiq/copay_notifications/parse_new_statements_job.rb:20]: I believe there is no PII. Specifically backtraces only include information regarding the file name, class name and function name and look like this:
```
./app/sidekiq/copay_notifications/new_statement_notification_job.rb:34:in `block in <class:NewStatementNotificationJob>'
./spec/sidekiq/copay_notifications/new_statement_notification_job_spec.rb:132:in `block (4 levels) in <top (required)>'
/Users/.rvm/gems/ruby-3.3.3/gems/webmock-3.23.1/lib/webmock/rspec.rb:39:in `block (2 levels) in <top (required)>'
/Users/.rvm/gems/ruby-3.3.3/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:124:in `block in run'
/Users/.rvm/gems/ruby-3.3.3/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:110:in `run'
/Users/.rvm/gems/ruby-3.3.3/gems/rspec-retry-0.6.2/lib/rspec_ext/rspec_ext.rb:12:in `run_with_retry'
/Users/.rvm/gems/ruby-3.3.3/gems/rspec-retry-0.6.2/lib/rspec/retry.rb:37:in `block (2 levels) in setup'
```
The platform also logs backtraces, for example [here](https://github.com/department-of-veterans-affairs/vets-api/blob/master/modules/simple_forms_api/app/services/simple_forms_api/s3/submission_builder.rb#L100-L102), so I think this is safe.

File: modules/debts_api/app/sidekiq/debts_api/v0/form5655/vha_submission_job.rb (hunk 5 above)
Comment (JunTaoLuo) [reply modules/debts_api/app/sidekiq/debts_api/v0/form5655/vha_submission_job.rb]: Oops left the comment out. When we took a look at this, the destroy method is an optimization for Redis cache which marks the entry as deleted and prevents persistence.

**Review by LindseySaari — APPROVED** [review]

**Pull request merged**